Modules/Microsoft365DSC/DSCResources/MSFT_IntuneMobileAppsMacOSLobApp/MSFT_IntuneMobileAppsMacOSLobApp.schema.mof

```diff
@@ -0,0 +1,32 @@
+[ClassVersion("1.0.0"), FriendlyName("IntuneMobileAppsMacOSLobApp")]
+class MSFT_IntuneMobileAppsMacOSLobApp : OMI_BaseResource
+{
+    [Key, Description("The unique identifier for an entity. Read-only. Inherited from mobileApp object.")] String Id;
+
+    [Write, Description("Optional Response Headers Variable.")] String ResponseHeadersVariable;
+    [Write, Description("Additional Parameters.")] Hashtable AdditionalProperties; #TODOK
+
+    [Write, Description("The description of the app. Inherited from mobileApp.")] String Description;
+    [Write, Description("Name of the device health script publisher")] String Developer;
+    [Write, Description("The admin provided or imported title of the app. Inherited from mobileApp.")] String DisplayName;
+    [Write, Description("Name of the device health script publisher")] String InformationUrl;
+    [Write, Description("The value indicating whether the app is marked as featured by the admin. Inherited from mobileApp.")] String IsFeatured;
+    [Write, Description("The large icon to be displayed in the app details and used for upload of the icon. Inherited from mobileApp.")] String LargeIcon; #TODOK
+    [Write, Description("Notes for the app. Inherited from mobileApp.")] String Notes;
+    [Write, Description("The owner of the app. Inherited from mobileApp.")] String Owner;
+    [Write, Description("The privacy statement Url. Inherited from mobileApp.")] String PrivacyInformationUrl;
+    [Write, Description("The publisher of the app. Inherited from mobileApp.")] String Publisher;
+    [Write, Description("The publishing state for the app. The app cannot be assigned unless the app is published. Inherited from mobileApp. Possible values are: notPublished, processing, published.")] MobileAppPublishingState PublishingState; #TODOK
+
+    [Write, Description("The list of group assignments for this mobile app. To construct, see NOTES section for ASSIGNMENTS properties and create a hash table.")] String Assignments[]; #TODOK
+    [Write, Description("The list of categories for this app. To construct, see NOTES section for CATEGORIES properties and create a hash table.")] String Categories[]; #TODOK
```

Comment:
Can you please share a link of code example about how to handle a list of strings while the type is enum?

Reply:
I didn't mean the `PublishingState`, I meant the `Categories` property. Don't want any confusion there. Instead of saving the categories with its Id and the display name, I believe it's "better" for us users of the module to only have the display name exported, then looked up by the resource and handled accordingly.

Reply:
ok, I will look into this.
In that case, as a user does it make sense to have to multiple scripts (.pms1's, mof's, examples, tests etc ...) for different types of Apps like Android, Windows, iOS, MacOS etc.? I noticed that they use the 4 common cmdlets as below:
New-MgBetaDeviceAppManagementMobileApp, Get-MgBetaDeviceAppManagementMobileApp, Update-MgBetaDeviceAppManagementMobileApp, Remove-MgBetaDeviceAppManagementMobileApp

Given the same cmdlets, I don't see any reason to create multiple files for different app types but I would like to know your thoughts.

Reply:
As it couldn't be said better: It depends 🥲 Depends on what you see fit. Do you think you can manage all the different app types with one single resource? With all types, e.g. MSI, LoB, Store etc? Then yes, you can. But you have to be aware of the nuances each resource might have, different properties, different fetches, content upload etc. If I'm honest with you, I don't think it's a good way to have everything in one single resource. What you can probably do is combine some things into the same resource. I'll let you decide how to handle it.

Reply:
Ok, that makes sense that handling  different properties, different fetches, content upload etc. maybe different for different apps types. But I was wondering why the Intune Apps' PS cmdlets (New-MgBetaDeviceAppManagementMobileApp, Get-MgBetaDeviceAppManagementMobileApp, Update-MgBetaDeviceAppManagementMobileApp, Remove-MgBetaDeviceAppManagementMobileApp)  do not have those different properties mentioned in the docs here? https://learn.microsoft.com/en-us/powershell/module/microsoft.graph.beta.devices.corporatemanagement/new-mgbetadeviceappmanagementmobileapp?view=graph-powershell-beta

While I can see the different properties in the Graph API docs as below:

1. Windows (Win32) apps: https://learn.microsoft.com/en-us/graph/api/intune-apps-win32lobapp-create?view=graph-rest-1.0&tabs=http - like it has: applicableArchitectures, installCommandLine, minimumCpuSpeedInMHz and many other properties that other apps types do not have.
2. macOS app: https://learn.microsoft.com/en-us/graph/api/intune-apps-macoslobapp-create?view=graph-rest-1.0&tabs=http - it has: bundleId, md5HashChunkSize etc... that other app types do not have.

Do you know why the PS cmdlet supports only the common params? If want to handle the params/properties specific to app type then how to do it? I will need to use Invoke-MgGraphRequest instead of the PS cmdlets that I am currently using in this .psm1 script?

Reply:
It's not that the PowerShell cmdlet "only" supports the common parameters. I rather think that in the `-BodyParameter` you could specify all properties necessary for the desired application type to be created (or with `-AdditionalProperties`). So I don't think that you would need to use `Invoke-MgGraphRequest` there, but I also didn't test it, so my answer is more a guess than a fact. 

If there are significant differences, it's by far easier to create separate resources for the app types and have all the necessary properties, including the `@odata.type` handling there, instead of including all properties of all types, then "dynamically" get the values, and if not all are provided, throw an error. That would be a horrendous user experience.

From my point of view, go with separate resources where necessary. If there are some that you might be able to combine, go for it. Otherwise, don't.

Reply:
I tried to add just one additional property specific to MacOS Lob app 'IgnoreVersionDetection', it's a Boolean property.
But it's not getting exported! I keep getting the below error. I've pushed that changes in this PR. This issue is not happening for one of the main Boolean property called "IsFeatured", I kept them both side-by-side, but I am not sure what's causing this error. There are a few app type specific properties that can be added for MacOS Lob app as "-AdditionalProperties". And many more for Win32 apps and other Lob app types. So I tried a few ways but none of them worked to fetch the additional properties :( 

1. Get-MgBetaDeviceAppManagementMobileAppAsMacOSLobApp -MobileAppId "8d027f94-0682-431e-97c1-827d1879fa79"
2.  Get-MgBetaDeviceAppManagementMobileApp -Filter "isof('microsoft.graph.macOSLobApp')"  -ExpandProperty "categories,assignments"  -Select *
3.  Get-MgBetaDeviceAppManagementMobileApp -Filter "isof('microsoft.graph.macOSLobApp')"  -ExpandProperty "categories,assignments"  -Select "id,displayName,IgnoreVersionDetection"
4. 2.  Get-MgBetaDeviceAppManagementMobileApp -All -Filter "isof('microsoft.graph.macOSLobApp')"  -ExpandProperty "categories,assignments" | Where-Object -FilterScript { $_.AdditionalProperties.'@odata.type' -eq '#microsoft.graph.macOSLobApp' }

The below one is coming while trying to Export the config:

System.Management.Automation.ParameterBindingArgumentTransformationException: Cannot process argument transformation on parameter 'IgnoreVersionDetection'. Cannot convert value "" to type "System.Boolean". Boolean parameters accept only Boolean values and numbers, such as $True, $False, 1 or 0. ---> System.Management.Automation.ArgumentTransformationMetadataException: Cannot convert value "" to type "System.Boolean". Boolean parameters accept only Boolean values and numbers, such as $True, $False, 1 or 0. ---> System.Management.Automation.PSInvalidCastException: Cannot convert value "" to type "System.Boolean". Boolean parameters accept only Boolean values and numbers, such as $True, $False, 1 or 0.
   at System.Management.Automation.ArgumentTypeConverterAttribute.ThrowPSInvalidBooleanArgumentCastException(Type resultType, Type convertType)
   at System.Management.Automation.ArgumentTypeConverterAttribute.CheckBoolValue(Object value, Type boolType)
   at System.Management.Automation.ArgumentTypeConverterAttribute.Transform(EngineIntrinsics engineIntrinsics, Object inputData, Boolean bindingParameters, Boolean bindingScriptCmdlet)
   --- End of inner exception stack trace ---
   at System.Management.Automation.ArgumentTypeConverterAttribute.Transform(EngineIntrinsics engineIntrinsics, Object inputData, Boolean bindingParameters, Boolean bindingScriptCmdlet)
   at System.Management.Automation.ParameterBinderBase.BindParameter(CommandParameterInternal parameter, CompiledCommandParameter parameterMetadata, ParameterBindingFlags flags)
   --- End of inner exception stack trace ---
   at System.Management.Automation.ExceptionHandlingOps.CheckActionPreference(FunctionContext funcContext, Exception exception)
   at System.Management.Automation.Interpreter.ActionCallInstruction`2.Run(InterpretedFrame frame)
   at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)
   at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)
"Error during Export:"
at Export-TargetResource, C:\Users\kajalp\Documents\GitHub\Microsoft365DSC\Modules\Microsoft365DSC\DSCResources\MSFT_IntuneMobileAppsMacOSLobApp\MSFT_IntuneMobileAppsMacOSLobApp.psm1: line 655
at Start-M365DSCConfigurationExtract, C:\Users\kajalp\Documents\GitHub\Microsoft365DSC\Modules\Microsoft365DSC\Modules\M365DSCReverse.psm1: line 682
at Export-M365DSCConfiguration, C:\Users\kajalp\Documents\GitHub\Microsoft365DSC\Modules\Microsoft365DSC\Modules\M365DSCUtil.psm1: line 1460
at <ScriptBlock>, <No file>: line 1

Reply:
I tried with a string property as well, it doesn't throw any error but it doesn't get exported either.
Note: Since the Additional Properties are not working, currently the params that are added are common to all apps as they come from the base type graph entity class: MobileApp

Reply:
@Kajalp1079 Accessing the properties inside of `AdditionalProperties` is in fact a case-sensitive matter. It doesn't work with `IgnoreVersionDetection`, but it does with `ignoreVersionDetection` (lower-case i). Please try it like that.

Reply:
I have been trying from cmdline only, that's why I mentioned cmdlets above, let me paste them again as below. But the main thing that I am not sure about is the Get-MgBetaDeviceAppManagementMobileApp  cmdlet does not have -AdditionalProperty as switch. So when you say run the cmdlet with lower case you mean in -Select ?

Get-MgBetaDeviceAppManagementMobileAppAsMacOSLobApp -MobileAppId "8d027f94-0682-431e-97c1-827d1879fa79"
Get-MgBetaDeviceAppManagementMobileApp -Filter "isof('microsoft.graph.macOSLobApp')" -ExpandProperty "categories,assignments" -Select *
Get-MgBetaDeviceAppManagementMobileApp -Filter "isof('microsoft.graph.macOSLobApp')" -ExpandProperty "categories,assignments" -Select "id,displayName,ignoreVersionDetection"
Get-MgBetaDeviceAppManagementMobileApp -All -Filter "isof('microsoft.graph.macOSLobApp')" -ExpandProperty "categories,assignments" | Where-Object -FilterScript { $_.AdditionalProperties.'@odata.type' -eq '#microsoft.graph.macOSLobApp' }

Reply:
@Kajalp1079 Yes, that's what I was referring to. Dictionaries can be case-sensitive, and `AdditionalProperties` is such a thing unfortunately. Hope that helps. 

```powershell
> $mobileApp = Get-MgbetaDeviceAppManagementMobileApp -Filter "isof('microsoft.graph.macOSLobApp')"
> $mobileApp.AdditionalProperties.IgnoreVersionDetection # Outputs nothing
> $mobileApp.AdditionalProperties.ignoreVersionDetection
True
```

Reply:
Got it, it worked for a primitive type i.e ignoreVersionDetection boolean property.
Next, I picked an array of complex type (which has 3 strings) called childApps. That got exported fine (I think). 
but the   .\M365TenantConfig.ps1 is throwing error as below.  I am not sure how to debug the .\M365TenantConfig.ps1 , so I have pushed the exported file temporarily in the latest iteration. All childApps related changes are also pushed. Can you please take a look at it?


At E:\DSC\5\M365TenantConfig.ps1:17 char:9
+         IntuneMobileAppsMacOSLobApp "IntuneMobileAppsMacOSLobApp-Skyp ...
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~
Undefined DSC resource 'IntuneMobileAppsMacOSLobApp'. Use Import-DSCResource to import the resource.
    + CategoryInfo          : ParserError: (:) [], ParseException
    + FullyQualifiedErrorId : ResourceNotDefined